src/django_upgrade/fixers/request_user_is_authenticated .py

```diff
@@ -0,0 +1,67 @@
+"""
+Update use of request.user.is_authenticated() to use request.user.is_authenticated
+https://docs.djangoproject.com/en/1.10/releases/1.10/#using-user-is-authenticated-and-user-is-anonymous-as-methods
+"""
+from __future__ import annotations
+
+import ast
+import sys
+from functools import partial
+from typing import Iterable
+
+from tokenize_rt import Offset
+from tokenize_rt import Token
+
+from django_upgrade.ast import ast_start_offset
+from django_upgrade.data import Fixer
+from django_upgrade.data import State
+from django_upgrade.data import TokenFunc
+from django_upgrade.tokens import find
+from django_upgrade.tokens import NAME
+from django_upgrade.tokens import OP
+
+
+fixer = Fixer(
+    __name__,
+    min_version=(1, 10),
+)
+
+
+@fixer.register(ast.Call)
+def visit_Call(
+    state: State,
+    node: ast.Call,
+    parents: list[ast.AST],
+) -> Iterable[tuple[Offset, TokenFunc]]:
+    if (
+        isinstance(node.func, ast.Attribute)
+        and node.func.attr == "is_authenticated"
+        and is_request_user_or_self_request_user(node.func.value)
+        and len(node.args) == 0
+    ):
+        yield (
+            ast_start_offset(node),
+            partial(rewrite_user_is_auth),
+        )
+
+
+def is_request_user_or_self_request_user(node: ast.AST) -> bool:
+    return (
+        isinstance(node, ast.Attribute)
+        and node.attr == "user"
+        and (
+            (isinstance(node.value, ast.Name) and node.value.id == "request")
+            or (
+                isinstance(node.value, ast.Attribute)
+                and isinstance(node.value.value, ast.Name)
+                and node.value.value.id == "self"
+                and node.value.attr == "request"
+            )
+        )
+    )
+
+
+def rewrite_user_is_auth(tokens: list[Token], i: int) -> None:
+    j = find(tokens, i, name=NAME, src="is_authenticated")
+    y = find(tokens, i, name=OP, src=")")
+    del tokens[j + 1 : y + 1]
```

Comment:
Find the opening parenthesis and delete from the opening to closing only, rather than `j+1`. There may be space, even newlines and comments, before the parenthesis!

Add these tests to see what I mean:

```py
request . user . is_authenticated ( )
```


```py
if (
    request
    .user
    .is_authenticated  # bla
    ()
):
    ...
```

Reply:
Thanks for the feedback, I also found a whitespace in the fixer's filename that I removed.

I am working on a fix for the additional cases you suggested but I encountered an issue.
I've changed the rewrite function like this:
```python
def rewrite_user_is_auth(tokens: list[Token], i: int) -> None:
    j = find(tokens, i, name=NAME, src="is_authenticated")
    y = find(tokens, j, name=OP, src="(")
    z = find(tokens, y, name=OP, src=")")
    del tokens[z]
    del tokens[y]
```
but I can't make it work with this case
```python
def test_spaces_and_comments():
    check_transformed(
        """\
        if (
            request
            .user
            .is_authenticated  # bla
            ()
        ):
            ...
        """,
        """\
        if (
            request
            .user
            .is_authenticated  # bla
            
        ):
            ...
        """,
        settings,
    )

```
This is the output of pytest
```shell
>       assert fixed == dedented_after
E       AssertionError: assert 'if (\n    request\n    .user\n    .is_authenticated  # bla\n    \n):\n    ...\n' == 'if (\n    request\n    .user\n    .is_authenticated  # bla\n    ()\n):\n    ...\n'
E
E           if (
E               request
E               .user
E               .is_authenticated  # bla
E         -     ()
E         ?     --
E         +     
E           ):
E               ...

tests\fixers\tools.py:25: AssertionError
```
It seems the only different between the two string are some whitespaces after the comment, so I tought it was because of my formatter but even disabling it or thinkering with the whitespaces didn't solve the issue.

I am sure I am missing something here, can you point me in the right direction?

Reply:
The problem is gonna be because either Black or `textwrap.dedent()` trims trailing whitespace in the file, whilst the fixer leaves it. We’re fine leaving the whitespace, django-upgrade leaves formatting to Black or whatever.

This version of the test should work:

```py
def test_spaces_and_comments():
    check_transformed(
        """\
        if (
            request
            .user
            .is_authenticated  # bla
            () # bla
        ):
            ...
        """,
        """\
        if (
            request
            .user
            .is_authenticated  # bla
             # bla
        ):
            ...
        """,
        settings,
    )
```

Reply:
Ah, doing two `del` operations will also break things. After deleting a token, the laters ones shift leftwards in position, decreasing their indexes. Best to delete a range instead - leaving a suggestion.

Reply:
I was deleting first the closing parenthesis w/ `del tokens[z]` and the the opening one w/ `del tokens[y]` to not shift positions. Doing as you suggest will remove the space between the parentheses in this case.
```python
def test_spaces_between():
    check_transformed(
        "request . user . is_authenticated ( )",
        "request . user . is_authenticated  ",
        settings,
    )
```
Is that fine?

Reply:
Yes, that is more consistent IMO. It will remove a comment in this obscure case:

```python
request.user.is_anonymous(  # something
)
```

But I think that’s fine.